### ISSUE TITLE
Question: Sticky Sessions

### ISSUE THREAD

**Issue by boustanihani** [andruschka/pm2-meteor#6]
AFAIK **PM2** does not support "Sticky Sessions" out-of-the-box.

How does **pm2-meteor** enable "Sticky Sessions" ??

**Comment by andruschka** [comment]
We use PM2s "cluster_mode"  and didn't experience any troubles yet. (Make sure to use nodejs >= v0.12) But you could also run in "fork_mode" and use nginx for load balancing. [take a look at socket.io docs](http://socket.io/docs/using-multiple-nodes/#nginx-configuration)

**Comment by boustanihani** [comment]
You mean PM2s "cluster_mode" supports "Sticky Sessions" already?

Have you checked this => https://github.com/Unitech/pm2/issues/389

**Comment by andruschka** [comment]
Oh ok I see the problem. Since we are trying to have a stateless server code - and saving shared vars- this didn't really come up. If you are in need of sticky sessions, you should use nginx as load ballancer.

**Comment by andruschka** [comment]
I will have a closer look at this tomorrow ;) Maybe I will add the sticky-session package to the building process

**Comment by boustanihani** [comment]
Cool :bowtie: :+1:

**Comment by boustanihani** [comment]
How about reopening this issue for the mean time :smiley:

**Comment by andruschka** [comment]
Yes I am working on it! This one isn't that simple. I will keep you updated!

**Comment by andruschka** [comment]
(The problem here is the Meteor.bindEnvironment stuff ...) I am currently testing sticky-session npm package

**Comment by andruschka** [comment]
Could you explain some use cases where PM2 cluster_mode and Meteor won't work.
Because if you follow the stateless server rules, it works like a charm! (DDP, calls, the whole Meteor stuff works in our production Apps - only Meteor.environmentVars have some problems...)

**Comment by boustanihani** [comment]
I still haven't tried this out...

Did you get any results with testing sticky-session npm package?

**Comment by andruschka** [comment]
I tried using that package with Meteor but unfortunately this isnt that simple. 
As we are saving all serverside data to the DB and using Nodejs Jobs we don't have any problems with this...

**Comment by hungchai** [comment]
how 's the progress??

**Comment by andruschka** [comment]
I will release a post this week on how to build / deploy Meteor apps the right way for PM2 (and other process manager). We are currently running about 5 bigger Meteor apps - without the need of sticky sessions. You just have to stick to some rules and you won't have a problem. I don't think that I will focus on this one ... contributions are welcome !

**Comment by andruschka** [comment]
Or if someone have some specific sticky session problems - please post them here !!!

**Comment by hungchai** [comment]
looking forward to your post

**Comment by hungchai** [comment]
I am still waiting your post :+1:

**Comment by andruschka** [comment]
sorry - moved to a new city this month. still working on it :)

**Comment by hungchai** [comment]
Finally, I use 

> meteorhacks/cluster

And

> pm2-meteor

 together 

to solve the problem. 

![capture1](https://cloud.githubusercontent.com/assets/10190523/19180541/2fae4074-8c99-11e6-85c0-bdf6d3343072.PNG)
![capture2](https://cloud.githubusercontent.com/assets/10190523/19180544/33f52396-8c99-11e6-9b7f-d9afe9f57874.PNG)

**Comment by dnish** [comment]
@andruschka I'm just wondering if the missing sticky session support will cause issues like this:

https://forums.meteor.com/t/app-stopped-working-but-it-is-online/33990/2

What would happen if POST request <code>xhr</code> goes to instance 1 and the next <code>xhr_send</code> to instance 2? Would instance 2 response with a 404 because the client asked for connection on instance 1?

If I take a look [here](http://stackoverflow.com/questions/23808528/sockjs-xhr-requests-404ing-with-meteor-app-running-behind-passenger-nginx), it seems to the that issue. We're having this one now in our production apps, both on pm2-meteor and some users who only see a loaded page without content (subs and methods were not called). I've just run into this issue this evening on my Mac machine. I don't know in which case this happens, but it happens and seems to be a problem on production apps.

// Edit: Just tried a simulation with disabled WebSockets (plugin needed for FF: https://addons.mozilla.org/de/firefox/addon/websocket-disabler/). If you disable WebSockets and you visit a deployed pm2-meteor application, your client will run full of <code>INVALID_STATE_ERR</code> errors.

**Comment by artpolikarpov** [comment]
@andruschka, sticky sessions is essential for scaling Meteor. But moslty it needed for sockjs fallback (when websockets not working). In that situation each new XHR-request from client will go to a random instance and all subscriptions will go crazy.

And stickiness should be IP-based (as in "sticky-session" package), but not cookie-based.

**Comment by hungchai** [comment]
@andruschka 
Do you mean if using Meteor with websocket disabled, it will get sticky sessions in pm2 cluster mode?

**Comment by elie222** [comment]
Nginx can handle the sticky sessions using fork mode:
https://github.com/andruschka/pm2-meteor/issues/55

**Comment by hungchai** [comment]
My solution is HaProxy + fork mode
